File: decorator.py
""" The Decorator Pattern attaches additional responsibilities to an object dynamically. Decorators provide a flexible alternative to subclassing for extending functionality.

----->Designing a hotel ordering system (add quanitity, extra cheese etc , eg zomato food ordering system)
----->Designing a Shopping cart ( With each user as Base decorator and items as extra)

"""

from abc import abstractclassmethod

#abstract beverage class
class BeverageAbstract:
    @abstractclassmethod
    def __init__(self):
        self.description="Unknown Beverage"
    

    @abstractclassmethod
    def getDescription(self):
        return self.description
    
    @abstractclassmethod
    def getCost(self):
        pass

#abstact Decorator the main decorator
class mainDecorator(BeverageAbstract):

    @abstractclassmethod
    def __init__(self):
        pass

    @abstractclassmethod
    def getDescription(self):
        pass

    @abstractclassmethod
    def getCost(self):
        pass


class Espresso(BeverageAbstract):

    @classmethod
    def __init__(self):
        self.description="Espresso"

    @classmethod
    def getCost(self):
        return 0.20

class HouseBlend(BeverageAbstract):

    @classmethod
    def __init__(self):
        self.description="House Blend"

    @classmethod
    def getCost(self):
        return 1.20

class Mocha(mainDecorator):

    @classmethod
    def __init__(self,beverage):
        self.beverage=beverage

    @classmethod
    def getDescription(self):
        return self.beverage.getDescription()+" mocha"
    
    @classmethod
    def getCost(self):
        return self.beverage.getCost()+0.20

class Whip(mainDecorator):

    @classmethod
    def __init__(self,beverage):
        self.beverage=beverage

    @classmethod
    def getDescription(self):
        return self.beverage.getDescription()+" whip"
    
    @classmethod
    def getCost(self):
        return self.beverage.getCost()+1.10

def  DecoratorTest():
    beverage=Espresso()
    print(beverage.getDescription(),beverage.getCost())
    beverage=Mocha(beverage)
    print(beverage.getDescription(),beverage.getCost())
    beverage=Whip(beverage)
    print(beverage.getDescription(),beverage.getCost())



DecoratorTest()

File: decorator.go
// decorator pattern in go
// What we need , we need 2 abstract classes one for main object second for decorator
// which extends the main object
package main

import "fmt"

type abstarctBeverage interface {
	getDescription() string
	getCost() float32
}

type coffee struct {
}

func (c *coffee) getDescription() string {
	description := "Espresso"
	return description
}

func (c *coffee) getCost() float32 {
	return 20.00
}

type mocha struct {
	coffee abstarctBeverage
}

func (m *mocha) getDescription() string {
	description := m.coffee.getDescription() + "Mocha"
	return description
}
func (m *mocha) getCost() float32 {
	return m.coffee.getCost() + 0.5
}

func main() {
	coffee := &coffee{}
	fmt.Println(coffee.getDescription())
	mocha := &mocha{
		coffee: coffee,
	}
	fmt.Println(mocha.getDescription())
}
